source/containerimage/pull.go

```diff
@@ -100,136 +101,171 @@ func (is *imageSource) Resolve(ctx context.Context, id source.Identifier, sm *se
+
 	pullerUtil := &pull.Puller{
-		Snapshotter:  is.Snapshotter,
 		ContentStore: is.ContentStore,
-		Applier:      is.Applier,
-		Src:          imageIdentifier.Reference,
-		Resolver:     pull.NewResolver(ctx, is.RegistryHosts, sm, is.ImageStore, imageIdentifier.ResolveMode, imageIdentifier.Reference.String()),
+		Resolver:     resolver,
 		Platform:     &platform,
+		Src:          imageIdentifier.Reference,
 	}
 	p := &puller{
 		CacheAccessor: is.CacheAccessor,
+		LeaseManager:  is.LeaseManager,
 		Puller:        pullerUtil,
-		Platform:      platform,
 		id:            imageIdentifier,
-		LeaseManager:  is.LeaseManager,
+		vtx:           vtx,
 	}
 	return p, nil
 }
 
 type puller struct {
-	CacheAccessor cache.Accessor
-	LeaseManager  leases.Manager
-	Platform      specs.Platform
-	id            *source.ImageIdentifier
+	CacheAccessor    cache.Accessor
+	LeaseManager     leases.Manager
+	id               *source.ImageIdentifier
+	vtx              solver.Vertex
+	descHandler      *cache.DescHandler
+	manifest         *pull.PulledManifests
+	releaseTmpLeases func(context.Context) error
 	*pull.Puller
 }
 
-func mainManifestKey(ctx context.Context, desc specs.Descriptor, platform specs.Platform) (digest.Digest, error) {
-	dt, err := json.Marshal(struct {
+func mainManifestKey(ctx context.Context, desc specs.Descriptor, platform *specs.Platform) (digest.Digest, error) {
+	keyStruct := struct {
 		Digest  digest.Digest
 		OS      string
 		Arch    string
 		Variant string `json:",omitempty"`
 	}{
-		Digest:  desc.Digest,
-		OS:      platform.OS,
-		Arch:    platform.Architecture,
-		Variant: platform.Variant,
-	})
+		Digest: desc.Digest,
+	}
+	if platform != nil {
+		keyStruct.OS = platform.OS
+		keyStruct.Arch = platform.Architecture
+		keyStruct.Variant = platform.Variant
+	}
+
+	dt, err := json.Marshal(keyStruct)
 	if err != nil {
 		return "", err
 	}
 	return digest.FromBytes(dt), nil
 }
 
-func (p *puller) CacheKey(ctx context.Context, index int) (string, bool, error) {
-	_, desc, err := p.Puller.Resolve(ctx)
+func (p *puller) CacheKey(ctx context.Context, index int) (_ string, cacheOpts solver.CacheOpts, _ bool, err error) {
+	ctx, done, err := leaseutil.WithLease(ctx, p.LeaseManager, leases.WithExpiration(5*time.Minute), leaseutil.MakeTemporary)
 	if err != nil {
-		return "", false, err
+		return "", nil, false, err
 	}
+	p.releaseTmpLeases = done
+	imageutil.AddLease(p.releaseTmpLeases)
+	defer func() {
+		if err != nil {
+			p.releaseTmpLeases(ctx)
+		}
+	}()
+
+	resolveProgressDone := oneOffProgress(ctx, "resolve "+p.Src.String())
+	defer func() {
+		resolveProgressDone(err)
+	}()
+
+	p.manifest, err = p.PullManifests(ctx)
+	if err != nil {
+		return "", nil, false, err
+	}
+
+	if len(p.manifest.Remote.Descriptors) > 0 {
+		topDesc := p.manifest.Remote.Descriptors[len(p.manifest.Remote.Descriptors)-1]
```

Comment:
My thinking here (which I should probably add a comment for if nothing else) is that while it's true a parent desc could be in the chain of multiple refs and thus come from multiple remotes, it actually won't matter.

Say here you have a chain of descriptors from parent->child of `A->B->C` (`topDesc` is thus `C`), but in a different image you end up with just the chain `A->B` (where `topDesc` is thus `B`) and use an entirely different remote (and progress writer, etc.). If you end up creating lazy-refs for both these desc chains, say then that the `A->B` chain gets de-lazied first.

When you later get around to de-lazying the `A->B->C` chain, the implementation in `cache` realizes that `A->B` was already de-lazied by something else and [should skip de-lazying those descriptors](https://github.com/sipsma/buildkit/blob/32e67c439cf08881a91053ce00d62be6bdf7c88c/cache/refs.go#L540-L542). It will end up just de-lazying what it needs to, `C` in this case. It thus shouldn't matter that there were some descriptors that had different remotes in different refs.

In addition, for a single ref, all the descriptors in the chain for that ref are going to have the same `DescHandler` values, so setting the `DescHandler` for each in the chain seemed unnecessary.

That's the theory anyways; you asked in some other comments about synchronization issues which I need to go look deeper at as obviously that's important to get right for all the above to work. Let me know if this makes sense or if I'm missing something.

Reply:
But what if you have `A->B->C` and `A->D->C` competing?

Reply:
I forgot that the digest of a layer descriptor doesn't include the digest of its parent (this is all still settling in my brain), but I actually still feel like it would still work. If `A->B->C` and `A->D->C` are getting de-lazied at the same, each of the individual descriptors will still get de-lazied just once (again, assuming the locking/synchronization is done correctly). It's possible to end up with a mix of like `A` and `C` got de-lazied using `remote 1` and `B` got de-lazied using `remote 2`, but at the end of the day they are the same bytes no matter the remote and end up in the local content store, which is what matters, right? What am I missing still?

Reply:
> A and C got de-lazied using remote 1 and B got de-lazied using remote 2, but at the end of the day they are the same bytes

I'm not following on the "but at the end of the day they are the same bytes". The case I'm thinking is when there are 2 separate lazy refs (different bytes) that have the same top layer. Afaics when they both get unlazy request only one deschandler is used. So ref that was based on `A->B->C` but is pulled with a deschandler for `A->D->C` there could be a problem.

Reply:
Right, makes total sense, I was still somehow getting hung up on the layer digest not including its parent digest, but I get it now. I'll change this to set the DescHandler for each desc in the chain, thanks for bearing with me